src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -316,6 +326,125 @@ fn suggest_restriction(
+                                        )
+                                        | (GenericArgKind::Const(_), GenericArgKind::Const(_)) => {
+                                            true
+                                        }
+                                        _ => false,
+                                    }
+                                },
+                            )
+                            && !matches!(trait_ref.self_ty().kind, ty::Infer(_))
+                        {
+                            turbofish_suggestions.insert(trait_ref.self_ty(), eval_result);
+                        }
+                    };
+                }
+            })
+        });
+        // Sort types by always suggesting `Vec<_>` and `String` first, as they are the
+        // most likely desired types. Otherwise sort first by `EvaluationResult` and then by their
+        // string representation.
+        let mut turbofish_suggestions = turbofish_suggestions.into_iter().collect::<Vec<_>>();
+        turbofish_suggestions.sort_by(|left, right| {
+            let vec_type = self.tcx.get_diagnostic_item(sym::vec_type);
+            let string_type = self.tcx.get_diagnostic_item(sym::string_type);
+            match (&left.0.kind, &right.0.kind) {
+                (
+                    ty::Adt(ty::AdtDef { did: left, .. }, _),
+                    ty::Adt(ty::AdtDef { did: right, .. }, _),
+                ) if left == right => Ordering::Equal,
+                (
+                    ty::Adt(ty::AdtDef { did: left, .. }, _),
+                    ty::Adt(ty::AdtDef { did: right, .. }, _),
+                ) if Some(*left) == vec_type && Some(*right) == string_type => Ordering::Less,
+                (
+                    ty::Adt(ty::AdtDef { did: left, .. }, _),
+                    ty::Adt(ty::AdtDef { did: right, .. }, _),
+                ) if Some(*right) == vec_type && Some(*left) == string_type => Ordering::Greater,
+                (ty::Adt(ty::AdtDef { did, .. }, _), _)
+                    if Some(*did) == vec_type || Some(*did) == string_type =>
+                {
+                    Ordering::Less
+                }
+                (_, ty::Adt(ty::AdtDef { did, .. }, _))
+                    if Some(*did) == vec_type || Some(*did) == string_type =>
+                {
+                    Ordering::Greater
+                }
+                _ => {
+                    // We give preferential place in the suggestion list to types that will apply
+                    // without doubt, to push types with abiguities (may or may not apply depending
+                    // on other obligations we don't have access to here) later in the sugg list.
+                    if left.1 == right.1 {
+                        left.0.to_string().cmp(&right.0.to_string())
+                    } else {
+                        left.1.cmp(&right.1)
+                    }
+                }
+            }
+        });
```

Comment:
@nikomatsakis This whole method is trying to find all the type that have `impl`s that would fit an un-inferred type param. For example for `foo.collect()` we need to suggest some type for the param `B`. This can be anything that `impl`s `IntoIterator`, but we need to verify that all the obligations are fulfilled. Currently we have one hack (line 362) to make sure that `Item` in `IntoIterator<Item = _>` matches the type we have and we are _not_ evaluating further bounds, which means that we suggest `PathBuf` even though that `impl` has a bound for `T: AsRef<Path>`.